Use npm ci when running JS tests

PVince81 opened the pull request:
This makes sure that the pinned versions are selected correctly.
Fixes the colors.js incident.

This should make CI for JS tests work again.

PVince81 committed:
Use npm ci when running JS tests

This makes sure that the pinned versions are selected correctly.
Fixes the colors.js incident.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
interesting... so the colors.js issue has disappeared, but now the errors on CI match what I always see locally.
I never understood how the tests could even pass on CI if they don't locally, probably an env discrepancy

so at least now the envs are aligned (at least for mine), so it looks like those issues will need solving also

CarlSchwan commented:
> interesting... so the colors.js issue has disappeared, but now the errors on CI match what I always see locally. I never understood how the tests could even pass on CI if they don't locally, probably an env discrepancy
> 
> so at least now the envs are aligned (at least for mine), so it looks like those issues will need solving also

We probably should update the dependencies and then pin the colors.js dependenfy to 1.4.0

PVince81 commented:
@CarlSchwan sounds like a good plan, I'll try that. Thanks

PVince81 committed:
Update node deps for JS tests

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I ran npm install with the pinned dep for colors and pushed the updated package-lock.json.
The same test errors still happen.

PVince81 commented:
I saw that recently more stuff got bundled into the JS files, even stuff that isn't supposed to auto-load. Causes some XHR to be done too early. Something to examine later.

I had a look at the contacts menu failure where 1 call to console.error is expected instead of 3.
The calls all look like this:
```
LOG LOG: '##### call 0: [Vue warn]: Cannot find element: #files-app-settings'
LOG LOG: '##### call 1: [Vue warn]: Cannot find element: #app-sidebar'
LOG LOG: '##### call 2: There was an error loading your contacts,[object Object]'
```

so possibly it's those additional bundles that get loaded that cause more side effects

PVince81 committed:
Fix JS tests after test lib updates

Prevent XHR during load by checking window.TESTING.
Adjust some expected values.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
fixed the tests, they pass locally with Chromium and Firefox

PVince81 commented on apps/files/src/main-personal-settings.js:
```diff
@@ -32,5 +32,7 @@ __webpack_nonce__ = btoa(getRequestToken())
 
-const View = Vue.extend(PersonalSettings)
-new View().$mount('#files-personal-settings')
+if (!window.TESTING) {
```
@artonge FYI I saw that you've added those bundles in karma.config.js, which now caused side effects because those bootstrap files were loaded and were attempting to interact with missing DOM elements or run XHR requests right on start.

artonge replied:
The failing tests were failing before I touched karma.config, didn't try to fix them as they were only failing locally.

PVince81 commented:
I've re-requested a review due to the changes being bigger.
Please have another look, thanks!

artonge reviewed (approved):
One step closer to making them work with npm 7 !

PVince81 commented:
/backport to stable23

PVince81 commented:
/backport to stable22

PVince81 commented:
/backport to stable21

PVince81 commented:
(I'm sure there will be backport conflicts, am already bracing for them...)

artonge commented:
> (I'm sure there will be backport conflicts, am already bracing for them...)

Maybe put build files into a separate commits so it eases backports

PVince81 commented:
- [ ] stable23: https://github.com/nextcloud/server/pull/30580
- [ ] stable22
- [ ] stable21
